### DISCUSSION TITLE
Failed Updates of MSI Setup Files (another version installed)

### DISCUSSION THREAD

**Discussion by maxsmartstep** [Romanitho/Winget-AutoUpdate#249] [Q&A]
Hi, thanks for this cool powershell scripts, helps a lot in my dayly bussines.

Actually i have a problem of automatic updating with MSI Setup files, i getting errors like "another version of this Package ist already installed"

In my enviroment the error is comming up with apps like "Cisco.CiscoWebexMeetings" or "PuTTY.PuTTY"
The apps where also installed via intune with the script "Winget-Install"

Does anybody hase the same issue or a workaround for it?
Is there the option to uninstall the old version before or to force the installation?
On internet research i found something for MS-Teams: [GitHub - microsoft/TeamsMsiOverride: Scripts to simplify the updating of the Teams Machine-Wide Installer, and allow clients to update from the MSI.](https://github.com/microsoft/TeamsMsiOverride)

**Comment by KnifMelti** [comment]
(EDIT: `-preinstall` doesn't work in this scenario uninstalling before upgrading (then the upgrade doesn't find the product)!)
EDIT2: `-preinstall` DO work if you add the install too afterwards in the script (see last line)!

If downloading **`Winget-AutoUpdate-main.zip`** (no **Release** yet) from **Code** you could use the mods function `-preinstall`
When it comes to ex. **PuTTY.PuTTY** create a `PuTTY.PuTTY-preinstall.ps1` under` .\mods` to uninstall (if the former version is **msi**) the previous version.
Or edit the below script to your hearts desire ✌
```
$App = "PuTTY*"
$InstalledSoftware = Get-ChildItem "HKLM:\Software\Microsoft\Windows\CurrentVersion\Uninstall"
foreach ($obj in $InstalledSoftware){
	if ($obj.GetValue('DisplayName') -like $App) {
		$UninstallString = $obj.GetValue('UninstallString')
		if ($UninstallString -like "MsiExec.exe*") {
			$ProductCode = select-string "{.*}" -inputobject $UninstallString
			$ProductCode = $ProductCode.matches.groups[0].value
			Start-Process "C:\Windows\System32\msiexec.exe" -ArgumentList "/x$ProductCode REBOOT=R /qn" -Wait
		}
	}
}
& $Winget install --id PuTTY.PuTTY --accept-package-agreements --accept-source-agreements -h
```

**Reply by KnifMelti** [reply]
...but testing at home I can't even uninstall the current version of PuTTY 0.78 (MSI error says 1605 - the product isn't installed) from Control Panel etc.. ..so something is definitely wrong with the package - not so uncommon in **Winget Land**...

**Reply by Romanitho** [reply]
Putty was buggy the way it get updated on the v0.78 if I remeber well. I had to uninstall it manually before reinstalling. Don't know for Cisco

**Comment by KnifMelti** [comment]
`[FINAL EDIT]` 😁 Now a total managed automatic uninstall routine...
`[Last EDIT]` Now a total managed uninstall routine for everything...
Made a general uninstall routine (for common uninstallers) that can be used in `AppID-preinstall.ps1`/`AppID-preuninstall.ps1` mods:
```
#Beginning of AppName string (wildcard after)...
$App = "AppName*"

$InstalledSoftware = Get-ChildItem "HKLM:\Software\Microsoft\Windows\CurrentVersion\Uninstall"
foreach ($obj in $InstalledSoftware){
	if ($obj.GetValue('DisplayName') -like $App) {
		$UninstallString = $obj.GetValue('UninstallString')
		if ($UninstallString -like "MsiExec.exe*") {
			$ProductCode = Select-String "{.*}" -inputobject $UninstallString
			$ProductCode = $ProductCode.matches.groups[0].value
			#MSI x64 Installer
			$Exec = Start-Process "C:\Windows\System32\msiexec.exe" -ArgumentList "/x$ProductCode REBOOT=R /qn" -PassThru -Wait
			#Stop Hard Reboot (if bad MSI!)
			if ($Exec.ExitCode -eq 1641) {
				Start-Process "C:\Windows\System32\shutdown.exe" -ArgumentList "/a"
			}
		}
		else {
			$QuietUninstallString = $obj.GetValue('QuietUninstallString')
			if ($QuietUninstallString) {
				$QuietUninstallString = Select-String "(\x22.*\x22) +(.*)" -inputobject $QuietUninstallString
				$Command = $QuietUninstallString.matches.groups[1].value
				$Parameter = $QuietUninstallString.matches.groups[2].value
				#All EXE x64 Installers (already defined silent uninstall)
				Start-Process $Command -ArgumentList $Parameter -Wait
			}
			else {
				$NullSoft = Select-String -Path $UninstallString.Trim([char]0x0022) -Pattern "Nullsoft.NSIS"
				if ($NullSoft) {
					#NSIS x64 Installer
					Start-Process $UninstallString -ArgumentList "/S" -Wait
				}
				else {
					$Inno = Select-String -Path $UninstallString.Trim([char]0x0022) -Pattern "Inno Setup"
					if ($Inno) {
						#Inno x64 Installer
						Start-Process $UninstallString -ArgumentList "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /SP-" -Wait
					}
					else {
						Write-Host "x64 Uninstaller unknown..."
					}
				}
			}
		}
		$x64 = $true
		break
	}
}
if (!$x64) {
	$InstalledSoftware = Get-ChildItem "HKLM:\Software\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall"
	foreach ($obj in $InstalledSoftware){
		if ($obj.GetValue('DisplayName') -like $App) {
			$UninstallString = $obj.GetValue('UninstallString')
			if ($UninstallString -like "MsiExec.exe*") {
				$ProductCode = Select-String "{.*}" -inputobject $UninstallString
				$ProductCode = $ProductCode.matches.groups[0].value
				#MSI x86 Installer
				$Exec = Start-Process "C:\Windows\System32\msiexec.exe" -ArgumentList "/x$ProductCode REBOOT=R /qn" -PassThru -Wait
				#Stop Hard Reboot (if bad MSI!)
				if ($Exec.ExitCode -eq 1641) {
					Start-Process "C:\Windows\System32\shutdown.exe" -ArgumentList "/a"
				}
			}
			else {
				$QuietUninstallString = $obj.GetValue('QuietUninstallString')
				if ($QuietUninstallString) {
					$QuietUninstallString = Select-String "(\x22.*\x22) +(.*)" -inputobject $QuietUninstallString
					$Command = $QuietUninstallString.matches.groups[1].value
					$Parameter = $QuietUninstallString.matches.groups[2].value
					#All EXE x86 Installers (already defined silent uninstall)
					Start-Process $Command -ArgumentList $Parameter -Wait
				}
				else {
					$NullSoft = Select-String -Path $UninstallString.Trim([char]0x0022) -Pattern "Nullsoft.NSIS"
					if ($NullSoft) {
						#NSIS x86 Installer
						Start-Process $UninstallString -ArgumentList "/S" -Wait
					}
					else {
						$Inno = Select-String -Path $UninstallString.Trim([char]0x0022) -Pattern "Inno Setup"
						if ($Inno) {
							#Inno x86 Installer
							Start-Process $UninstallString -ArgumentList "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /SP-" -Wait
						}
						else {
							Write-Host "x86 Uninstaller unknown..."
						}
					}
				}
			}
			break
		}
	}
}
```
If `AppID-preinstall.ps1` then afterwards:

`& $Winget install --id AppID.AppID --accept-package-agreements --accept-source-agreements -h`

**Reply by KnifMelti** [reply]
#250

**Comment by maxsmartstep** [comment]
Hi there,

Happy new year and thanks for implementing this wonderfull feature so fast.

I testet this in my enviroment, it seems that sometimes this funktion works, sometimes not.

But first for verrification my way of implementing this:

	- Copy _AppID-template.ps1 and rename to "AppID-preinstall.ps1
	- Set $Proc and $App variable
	- Under extras added the line "& $Winget install --id PuTTY.PuTTY --accept-package-agreements --accept-source-agreements -h"
	- Copy "AppID-preinstall.ps1" to "AppID-upgrade.ps1"


This will work in my enviroment for Cisco.WebexTeams. But for Apps like PuTTY.PuTTY or Inkscape.Inkscape the application will be updated (when i checked installed Version at the installed application directory) but in the Script update is failed:

14:07:13 - Updating Inkscape from 1.1.2 to 1.2.1...
14:07:24 - ##########   WINGET UPGRADE PROCESS STARTS FOR APPLICATION ID 'Inkscape.Inkscape'   ##########
14:07:30 - Modifications for Inkscape.Inkscape before upgrade are being applied...
14:08:46 - -> Running: Winget upgrade --id Inkscape.Inkscape --accept-package-agreements --accept-source-agreements -h
   -\| 

  ██▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  1024 KB / 11.6 MB
  █████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  2.00 MB / 11.6 MB
  ███████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  3.00 MB / 11.6 MB
  ██████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  4.00 MB / 11.6 MB
  ████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  5.00 MB / 11.6 MB
  ███████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  6.00 MB / 11.6 MB
  █████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒  7.00 MB / 11.6 MB
  ████████████████████▒▒▒▒▒▒▒▒▒▒  8.00 MB / 11.6 MB
  ███████████████████████▒▒▒▒▒▒▒  9.00 MB / 11.6 MB
  █████████████████████████▒▒▒▒▒  10.0 MB / 11.6 MB
  ████████████████████████████▒▒  11.0 MB / 11.6 MB
  ██████████████████████████████  11.6 MB / 11.6 MB
                                                              
   - 
No applicable update found.
14:08:48 - Modifications for Inkscape.Inkscape during upgrade are being applied...
14:09:14 - -> An upgrade for Inkscape failed, now trying an install instead...
14:09:14 - -> Running: Winget install --id Inkscape.Inkscape --accept-package-agreements --accept-source-agreements -h
   -\| 

  ██▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  1024 KB / 11.6 MB
  █████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  2.00 MB / 11.6 MB
  ███████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  3.00 MB / 11.6 MB
  ██████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  4.00 MB / 11.6 MB
  ████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  5.00 MB / 11.6 MB
  ███████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  6.00 MB / 11.6 MB
  █████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒  7.00 MB / 11.6 MB
  ████████████████████▒▒▒▒▒▒▒▒▒▒  8.00 MB / 11.6 MB
  ███████████████████████▒▒▒▒▒▒▒  9.00 MB / 11.6 MB
  █████████████████████████▒▒▒▒▒  10.0 MB / 11.6 MB
  ████████████████████████████▒▒  11.0 MB / 11.6 MB
  ██████████████████████████████  11.6 MB / 11.6 MB
                                                              
Found Inkscape [Inkscape.Inkscape] Version 1.2.1
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://media.inkscape.org/dl/resources/file/inkscape-1.2.1_2022-07-14_9c6d41e410-x64_qPmYh2W.msi

                                                              

  ▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  0.00 B /  136 MB
  ██▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  10.0 MB /  136 MB
  ████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  22.0 MB /  136 MB
  ███████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  35.0 MB /  136 MB
  ██████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  48.0 MB /  136 MB
  █████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  61.0 MB /  136 MB
  ████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒  73.0 MB /  136 MB
  ██████████████████▒▒▒▒▒▒▒▒▒▒▒▒  85.0 MB /  136 MB
  █████████████████████▒▒▒▒▒▒▒▒▒  97.0 MB /  136 MB
  ███████████████████████▒▒▒▒▒▒▒   108 MB /  136 MB
  ██████████████████████████▒▒▒▒   120 MB /  136 MB
  █████████████████████████████▒   133 MB /  136 MB
  ██████████████████████████████   136 MB /  136 MB
Successfully verified installer hash
Starting package install...
   -\|/-\|/-\|/- 
Successfully installed
14:09:36 - Modifications for Inkscape.Inkscape during install are being applied...
14:09:37 - ##########   WINGET UPGRADE PROCESS FINISHED FOR APPLICATION ID 'Inkscape.Inkscape'   ##########
14:09:37 - Inkscape update failed.


When i rerun the script, the error still excist (and the script want to update the same application again)

So, when i check the installed software via Windows GUI, i see that only Inkscape 1.1.2 (old version) is avalible in the list
When i check "winget list" i have 2 entrys, one for the old and one for the new version

**Reply by KnifMelti** [reply]
@Work now - will try to understand what you're doing when @Home

**Reply by KnifMelti** [reply]
First and furthermost, this uninstall routine (via _AppID-template.ps1/_Mods-Functions.ps1) isn't a **WAU** part (as stated in the directory mods [README.md](https://github.com/Romanitho/Winget-AutoUpdate/pull/250/files#diff-3d88d1170ee8beb993cbf88bbbc38abddd64e6bd9cc841f72aaec1ce3b72ff3c) (_A script Template and Mods Functions are included **as example** to get you started..._)...
...it's up to the individual to write the scripts for the **mods** feature (I only included **my own logic** as scripts).
The **WAU** part is only that it now **also** detects and runs **AppID-preinstall.ps1** **before** the common upgrade/install way.

Secondly, in my latest **PR** #257 I changed the [_Mods-Functions.ps1](https://github.com/Romanitho/Winget-AutoUpdate/pull/250/files#diff-57ecebbd87526290b6c622fe021b32f0d45886a70303ffbcddba4d29a11c25ca) a little (i.e added support for **native defined uninstall** if not **MSI/Nullsoft/Inno**) - see down below.

Now, for your way of implementing it (if you have all the right files):
- Copy _AppID-template.ps1 and rename to "AppID-preinstall.ps1 (**check**, if the **AppID** is replaced by **Inkscape.Inkscape** or **PuTTY.PuTTY**)
- Set $Proc and $App variable (**check**, **$Proc** to kill and **$App** with a `*` at the end; search in PS is done with `-like`)
- Under extras added the line "& $Winget install --id PuTTY.PuTTY --accept-package-agreements --accept-source-agreements -h" (check, for installing **PuTTY** and if without `""`)
- Copy "AppID-preinstall.ps1" to "AppID-upgrade.ps1" (? - no go, **AppID-preinstall.ps1** should be the only one; it runs before **WAU** tries to upgrade/install **AppID** the common way)
**AppID-upgrade.ps1** runs **just after WAU** tries to upgrade **AppID** the common way, and you already have **AppID-preinstall.ps1** trying to manage these special cases.

As mentioned before: https://github.com/Romanitho/Winget-AutoUpdate/discussions/249#discussioncomment-4411813
It seems that the Product Code for **PuTTY** is wrong, ending in an MSI-error **1605**
...as said before: _A script Template and Mods Functions are included **as example** to get you started..._  

Latest **_Mods-Functions.ps1**:
```
#Common shared functions for mods handling

function Stop-ModsProc ($Proc) {
    foreach ($process in $Proc)
    {
        Stop-Process -Name $process -Force -ErrorAction SilentlyContinue | Out-Null
    }
    Return
}
function Wait-ModsProc ($Wait) {
    foreach ($process in $Wait)
    {
        Get-Process $process -ErrorAction SilentlyContinue | Foreach-Object { $_.WaitForExit() }
    }
    Return
}
function Uninstall-ModsApp ($App) {
    $InstalledSoftware = Get-ChildItem "HKLM:\Software\Microsoft\Windows\CurrentVersion\Uninstall"
    foreach ($obj in $InstalledSoftware){
        if ($obj.GetValue('DisplayName') -like $App) {
            $UninstallString = $obj.GetValue('UninstallString')
            $CleanedUninstallString = $UninstallString.Trim([char]0x0022)
            if ($UninstallString -like "MsiExec.exe*") {
                $ProductCode = Select-String "{.*}" -inputobject $UninstallString
                $ProductCode = $ProductCode.matches.groups[0].value
                #MSI x64 Installer
                $Exec = Start-Process "C:\Windows\System32\msiexec.exe" -ArgumentList "/x$ProductCode REBOOT=R /qn" -PassThru -Wait
                #Stop Hard Reboot (if bad MSI!)
                if ($Exec.ExitCode -eq 1641) {
                    Start-Process "C:\Windows\System32\shutdown.exe" -ArgumentList "/a"
                }
            }
            else {
                $QuietUninstallString = $obj.GetValue('QuietUninstallString')
                if ($QuietUninstallString) {
                    $QuietUninstallString = Select-String "(\x22.*\x22) +(.*)" -inputobject $QuietUninstallString
                    $Command = $QuietUninstallString.matches.groups[1].value
                    $Parameter = $QuietUninstallString.matches.groups[2].value
                    #All EXE x64 Installers (already defined silent uninstall)
                    Start-Process $Command -ArgumentList $Parameter -Wait
                }
                else {
                    if ((Test-Path $CleanedUninstallString)) {
                        $NullSoft = Select-String -Path $CleanedUninstallString -Pattern "Nullsoft"
                    }
                    if ($NullSoft) {
                        #NSIS x64 Installer
                        Start-Process $UninstallString -ArgumentList "/S" -Wait
                    }
                    else {
                        if ((Test-Path $CleanedUninstallString)) {
                            $Inno = Select-String -Path $CleanedUninstallString -Pattern "Inno Setup"
                        }
                        if ($Inno) {
                            #Inno x64 Installer
                            Start-Process $UninstallString -ArgumentList "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /SP-" -Wait
                        }
                        else {
                            Write-Host "x64 Uninstaller unknown, trying the UninstallString from registry..."
                            $NativeUninstallString = Select-String "(\x22.*\x22) +(.*)" -inputobject $UninstallString
                            $Command = $NativeUninstallString.matches.groups[1].value
                            $Parameter = $NativeUninstallString.matches.groups[2].value
                            #All EXE x64 Installers (native defined uninstall)
                            Start-Process $Command -ArgumentList $Parameter -Wait
                        }
                    }
                }
            }
            $x64 = $true
            break
        }
    }
    if (!$x64) {
        $InstalledSoftware = Get-ChildItem "HKLM:\Software\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall"
        foreach ($obj in $InstalledSoftware){
            if ($obj.GetValue('DisplayName') -like $App) {
                $UninstallString = $obj.GetValue('UninstallString')
                $CleanedUninstallString = $UninstallString.Trim([char]0x0022)
                if ($UninstallString -like "MsiExec.exe*") {
                    $ProductCode = Select-String "{.*}" -inputobject $UninstallString
                    $ProductCode = $ProductCode.matches.groups[0].value
                    #MSI x86 Installer
                    $Exec = Start-Process "C:\Windows\System32\msiexec.exe" -ArgumentList "/x$ProductCode REBOOT=R /qn" -PassThru -Wait
                    #Stop Hard Reboot (if bad MSI!)
                    if ($Exec.ExitCode -eq 1641) {
                        Start-Process "C:\Windows\System32\shutdown.exe" -ArgumentList "/a"
                    }
                }
                else {
                    $QuietUninstallString = $obj.GetValue('QuietUninstallString')
                    if ($QuietUninstallString) {
                        $QuietUninstallString = Select-String "(\x22.*\x22) +(.*)" -inputobject $QuietUninstallString
                        $Command = $QuietUninstallString.matches.groups[1].value
                        $Parameter = $QuietUninstallString.matches.groups[2].value
                        #All EXE x86 Installers (already defined silent uninstall)
                        Start-Process $Command -ArgumentList $Parameter -Wait
                    }
                    else {
                        if ((Test-Path $CleanedUninstallString)) {
                            $NullSoft = Select-String -Path $CleanedUninstallString -Pattern "Nullsoft"
                        }
                        if ($NullSoft) {
                            #NSIS x86 Installer
                            Start-Process $UninstallString -ArgumentList "/S" -Wait
                        }
                        else {
                            if ((Test-Path $CleanedUninstallString)) {
                                $Inno = Select-String -Path $CleanedUninstallString -Pattern "Inno Setup"
                            }
                            if ($Inno) {
                                #Inno x86 Installer
                                Start-Process $UninstallString -ArgumentList "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /SP-" -Wait
                            }
                            else {
                                Write-Host "x86 Uninstaller unknown, trying the UninstallString from registry..."
                                $NativeUninstallString = Select-String "(\x22.*\x22) +(.*)" -inputobject $UninstallString
                                $Command = $NativeUninstallString.matches.groups[1].value
                                $Parameter = $NativeUninstallString.matches.groups[2].value
                                #All EXE x86 Installers (native defined uninstall)
                                Start-Process $Command -ArgumentList $Parameter -Wait
                            }
                        }
                    }
                }
                break
            }
        }
    }
    Return
}
function Remove-ModsLnk ($Lnk) {
    foreach ($link in $Lnk)
    {
        Remove-Item -Path "${env:Public}\Desktop\$link.lnk" -Force -ErrorAction SilentlyContinue | Out-Null
    }
    Return
}

```

**Reply by KnifMelti** [reply]
An option when it comes to mismatched **MSI** installations/**Product Codes** on a larger scale is to use the old (not found at MS anymore) **command tool MsiZap.exe** that can be downloaded here: https://www.softpedia.com/get/Security/Secure-cleaning/Windows-Installer-CleanUp-Utility.shtml (after installing **msicuu2.exe** in admin context you find the command tool under `C:\Program Files (x86)\Windows Installer Clean Up`
To use it: https://support.microfocus.com/kb/doc.php?id=7023386#
With that command tool you can delete all registry values conflicting with a MSI installation

The problem with **PuTTY** is that 0.78 when installed didn't remove the MSI parts of registry for 0.77, resulting in two entries in registry:
![image](https://user-images.githubusercontent.com/102996177/210294688-818e2df9-a20f-4c61-8cd1-9b9579ecb918.png)

After that problem got detected they changed the manifest from **UpgradeBehavior: install** to **UpgradeBehavior: uninstallPrevious**: https://github.com/microsoft/winget-pkgs/tree/master/manifests/p/PuTTY/PuTTY

But, it was too late for most of us, 0.78 had already been installed, leaving two entries in registry.
**[EDIT]**: This doesn't work:
What you can do is:
- delete `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{E078C644-A120-4668-AD62-02E9FD530190}` (for PuTTY 0.77)

This you can automatically do in an **AppID-preinstall.ps1** under EXTRAS, nothing more in that **mod**!**[/EDIT]**

**Reply by KnifMelti** [reply]
There's more to `PuTTY.PuTTY` than meets the eye...
...in a **PuTTY.PuTTY-installed.ps1** you must now copy:
`C:\Windows\System32\config\systemprofile\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\PuTTY (64-bit)`
To:
`C:\ProgramData\Microsoft\Windows\Start Menu\Programs`

That's because it installs under **System** context but places the shortcuts totally wrong...
...and when uninstalling in a **PuTTY.PuTTY-preuninstall.ps1**:
Search for **PuTTY** in the registry uninstall and Uninstall the MSI/delete the directory:
`C:\ProgramData\Microsoft\Windows\Start Menu\Programs\PuTTY (64-bit)`

All this can be accomplished by using the files **_AppID-template.ps1**/**_Mods-Functions.ps1** in my latest PR #257 
**Winget Land**...
#259
